ext/nnstreamer/tensor_filter/tensor_filter_ncnn.cc

```diff
@@ -0,0 +1,503 @@
+    net.opt.use_vulkan_compute = true;
+    g_message ("accl = gpu\n");
+  } else {
+    net.opt.use_vulkan_compute = false;
+  }
+
+  // load model files
+  // ncnn returns nonzero value when an error occurs
+  if (prop->num_models == 1) {
+    if (net.load_param_bin (prop->model_files[0]))
+      throw std::invalid_argument (
+          "Failed to open the model file " + std::string (prop->model_files[0]));
+  } else {
+    if (net.load_param (prop->model_files[0]))
+      throw std::invalid_argument (
+          "Failed to open the param file " + std::string (prop->model_files[0]));
+    if (net.load_model (prop->model_files[1]))
+      throw std::invalid_argument (
+          "Failed to open the bin file " + std::string (prop->model_files[1]));
+  }
+
+  // get input layers from the ncnn network
+  const std::vector<int> &input_indexes = net.input_indexes ();
+  input_mats.clear ();
+  if (inputInfo.num_tensors != input_indexes.size ())
+    throw std::invalid_argument (
+        std::string ("Wrong number of input matrices")
+        + ": Found in argument = " + std::to_string (inputInfo.num_tensors)
+        + ", Found in model file = " + std::to_string (input_indexes.size ()));
+
+  // init input matrices
+  for (guint i = 0; i < inputInfo.num_tensors; i++) {
+    // get dimensions of the input matrix from inputInfo
+    const uint32_t *dim = gst_tensors_info_get_nth_info (&inputInfo, i)->dimension;
+    std::vector<int> shape;
+    while (*dim)
+      shape.push_back (*dim++);
+
+    // make ncnn matrix object
+    ncnn::Mat in;
+    switch (shape.size ()) {
+      case 1:
+        in = ncnn::Mat (shape[0]);
+        break;
+      case 2:
+        in = ncnn::Mat (shape[0], shape[1]);
+        break;
+      case 3:
+        in = ncnn::Mat (shape[0], shape[1], shape[2]);
+        break;
+      case 4:
+        in = ncnn::Mat (shape[0], shape[1], shape[2], shape[3]);
+        break;
+      default:
+        throw std::invalid_argument ("ncnn subplugin supports only up to 4 ranks and does not support input tensors of "
+                                     + std::to_string (shape.size ()) + " dimensions.");
+    }
+    input_mats.push_back (in);
+  }
+
+  // get output layers from the ncnn network
+  const std::vector<int> &output_indexes = net.output_indexes ();
+  output_mats.clear ();
+  if (outputInfo.num_tensors != output_indexes.size ())
+    throw std::invalid_argument (
+        std::string ("Wrong number of output matrices")
+        + ": Found in argument = " + std::to_string (outputInfo.num_tensors)
+        + ", Found in model file = " + std::to_string (output_indexes.size ()));
+
+  // init output matrices
+  output_mats.resize (outputInfo.num_tensors);
+
+  empty_model = false;
+}
+
+/**
+ * @brief Invoke ncnn model and get the inference result.
+ */
+void
+ncnn_subplugin::invoke (const GstTensorMemory *input, GstTensorMemory *output)
+{
+  if (empty_model)
+    throw std::runtime_error ("Model is empty.");
```

Comment:
Error message should be useful for those who may read it.
```
Model is empty: the ncnn instance is not configured and its "invoke" method is called. This may be an internal bug of nnstreamer or ncnn-subplugin unless if you have directly accessed ncnn-subplugin.
```

Reply:
Done.